### PR TITLE
WIP: Switch to containerd client

### PR DESCRIPTION
**Pull request by mikebrow**
This PR moves us to the latest containerd release. The major changes in containerd are implementation of a client library, removal of the rootfs service, renaming containers to tasks, ... 

Will address issue #49
 
First two commits are just to update godeps and get the code to compile.

TODO: will need a fake container client

Subsequent commits will be needed to update to client and get the code working again...

### CODE REVIEW

File: Godeps/Godeps.json [hunk 1]
```diff
@@ -387,64 +425,99 @@
...
+			"ImportPath": "golang.org/x/text/unicode/bidi",
+			"Rev": "19e51611da83d6be54ddafce4a4af510cb3e9ea4"
+		},
+		{
+			"ImportPath": "golang.org/x/text/unicode/norm",
+			"Rev": "19e51611da83d6be54ddafce4a4af510cb3e9ea4"
+		},
+		{
+			"ImportPath": "google.golang.org/genproto/googleapis/rpc/status",
+			"Rev": "aa2eb687b4d3e17154372564ad8d6bf11c3cf21f"
```
Comment (Random-Liu) [inline Godeps/Godeps.json]: d80a6e20e776b0b17a324d0ba1ab50a39c8e8944

Comment (mikebrow) [inline Godeps/Godeps.json]: thx

File: pkg/server/helpers.go [hunk 2]
```diff
@@ -320,9 +319,9 @@ func (c *criContainerdService) localResolve(ctx context.Context, ref string) (*m
 		}
 		image, err := c.imageStoreService.Get(ctx, normalized.String())
 		if err != nil {
-			if images.IsNotFound(err) {
-				return nil, nil
-			}
+			//if images.IsNotFound(err) { // TODO (mikebrow): IsNotFound removed
```
Comment (Random-Liu) [inline pkg/server/helpers.go]: Seems moved to https://github.com/containerd/containerd/blob/bdf9f5f7388e8203e63a74b89800f7f3dd4a7743/metadata/errors.go.

File: pkg/server/container_start.go [hunk 3]
```diff
@@ -172,24 +171,19 @@ func (c *criContainerdService) startContainer(ctx context.Context, id string, me
...
-			TypeUrl: runtimespec.Version,
-			Value:   rawSpec,
-		},
-		Rootfs:   mountsResp.Mounts,
-		Runtime:  defaultRuntime,
-		Stdin:    stdin,
-		Stdout:   stdout,
-		Stderr:   stderr,
-		Terminal: config.GetTty(),
+		ContainerID: id,
```
Comment (Random-Liu) [inline pkg/server/container_start.go]: Spec?

File: pkg/server/helpers.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/server/helpers.go]: Same with above.

File: pkg/server/image_remove.go [hunk 4]
```diff
@@ -55,7 +54,7 @@ func (c *criContainerdService) RemoveImage(ctx context.Context, r *runtime.Remov
 		// TODO(random-liu): Containerd should schedule a garbage collection immediately,
 		// and we may want to wait for the garbage collection to be over here.
 		err = c.imageStoreService.Delete(ctx, ref)
-		if err == nil || images.IsNotFound(err) {
+		if err == nil /*|| images.IsNotFound(err)*/ { // TODO(mikebrow): IsNotFound removed
```
Comment (Random-Liu) [inline pkg/server/image_remove.go]: Same.

File: pkg/server/sandbox_run.go [hunk 5]
```diff
@@ -138,19 +137,14 @@ func (c *criContainerdService) RunPodSandbox(ctx context.Context, r *runtime.Run
 	if err != nil {
 		return nil, fmt.Errorf("failed to generate sandbox container spec: %v", err)
 	}
-	rawSpec, err := json.Marshal(spec)
+	_, err = json.Marshal(spec)
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: ?

**Comment by Random-Liu** [comment]
@mikebrow Please switch the image part to use new client. And will do another review after that.

Let's try to finish this this week, I'll do review as soon as possible.

File: pkg/server/sandbox_run.go (hunk 5 above)
Comment (mikebrow) [inline pkg/server/sandbox_run.go]: unused.

**Comment by mikebrow** [comment]
```
@mikebrow Please switch the image part to use new client. And will do another review after that.
```
Yes, thats the plan first commits were just a mid step along the way.

File: pkg/server/container_start.go (hunk 3 above)
Comment (mikebrow) [inline pkg/server/container_start.go]: The spec is no longer the CreateRequest. I see a spec in the ExecRequest only.

**Comment by mikebrow** [comment]
Added TODO's regarding the requirement for using the new container client service to specify the spec on the container (associates the spec with a containerid). 

Comments addressed moving on to client changes.

**Comment by Random-Liu** [comment]
@mikebrow I think we should create/delete the container in this PR, or else this PR will not work.
Just something similar with https://github.com/containerd/containerd/blob/6428b4bad0c29e41a50b191d71969b5291451f6c/client.go#L172-L184.

For the unit, just create a empty interface, we'll get rid of it soon.

**Comment by mikebrow** [comment]
Yes, we have to use the new containers client service to make it work now. Though it seems to me it would be easier to just use the client than to do something similar to what the client is doing. Let's discuss.

**Comment by mikebrow** [comment]
rebased...

**Comment by Random-Liu** [comment]
Will take a look tonight.

**Comment by mikebrow** [comment]
Build server seems to have the hiccups tonight random timeouts on gomealinter

**Comment by mikebrow** [comment]
Important commits merged in #78 will close this PR and reopen a new ones for subsequent containerd/client updates

**Pull request closed without merging**